attest/verifier/src/ias.rs

```diff
@@ -0,0 +1,244 @@
+
+                        let existing_cert = signer_chain
+                            .last_mut()
+                            .expect("Somehow our per-signer chain was empty");
+                        let mut existing_cert_single = MbedtlsList::new();
+                        existing_cert_single.push(existing_cert.clone());
+
+                        if existing_cert.public_key_mut().write_public_der_vec()
+                            != chain_cert.public_key_mut().write_public_der_vec()
+                            && Certificate::verify_with_profile(
+                                &existing_cert_single,
+                                &chain_cert_single,
+                                None,
+                                Some(&profile),
+                                None,
+                            )
+                            .is_ok()
+                        {
+                            signer_chain.push(chain_cert);
+                            continue 'outer;
+                        }
+                    }
+
+                    break;
+                }
+                Some(signer_chain)
+            })
+            // Then see if any of those chains are connected to a trust anchor
+            .find_map(|mut signer_chain| {
+                let signer_toplevel = signer_chain
+                    .last()
+                    .expect("Signer chain was somehow emptied before use.");
+                let mut signer_toplevel_single = MbedtlsList::new();
+                signer_toplevel_single.push(signer_toplevel.clone());
+
+                // First, check if the last element in the chain is signed by a trust anchor
+                for cacert in &trust_anchors {
+                    let mut cacert_single = MbedtlsList::new();
+                    cacert_single.push(cacert.clone());
+                    if Certificate::verify_with_profile(
+                        &signer_toplevel_single,
+                        &cacert_single,
+                        None,
+                        Some(&profile),
+                        None,
+                    )
+                    .is_ok()
+                    {
+                        return Some(());
+                    }
+                }
+
+                // Otherwise, check if any of the pubkeys in the chain are a trust anchor
+                for cert in &mut signer_chain {
+                    for cacert in &mut trust_anchors {
+                        if cert.public_key_mut().write_public_der_vec()
+                            == cacert.public_key_mut().write_public_der_vec()
+                        {
+                            return Some(());
+                        }
+                    }
+                }
+                None
+            })
+            .ok_or(Error::BadSignature)?;
+
+        let report_data = VerificationReportData::try_from(report)?;
+
+        if (self.and_verifiers.is_empty()
+            || self
```

Comment:
I think this part that checks `is_empty` is unnecessary:
`.iter().all(...)` returns true if the set is empty, so we could delete the `is_empty` and the `||` part

https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.all

Reply:
Fair point. I'll leave it for now because Part 3 will end up replacing the `all()` usage here in favor of `filter_map().collect()`.